debezium-server/debezium-server-rocketmq/src/main/java/io/debezium/server/rocketmq/RocketMqChangeConsumer.java

```diff
@@ -0,0 +1,163 @@
+                            config.getValue(PROP_PRODUCER_SECRET_KEY, String.class)));
+        }
+        this.mqProducer = new DefaultMQProducer(rpcHook);
+        this.mqProducer.setNamesrvAddr(config.getValue(PROP_PRODUCER_NAME_SRV_ADDR, String.class));
+        this.mqProducer.setInstanceName(createUniqInstance(config.getValue(PROP_PRODUCER_NAME_SRV_ADDR, String.class)));
+        this.mqProducer.setProducerGroup(config.getValue(PROP_PRODUCER_GROUP, String.class));
+
+        if (config.getOptionalValue(PROP_PRODUCER_SEND_MSG_TIMEOUT, Integer.class).isPresent()) {
+            this.mqProducer.setSendMsgTimeout(config.getValue(PROP_PRODUCER_SEND_MSG_TIMEOUT, Integer.class));
+        }
+
+        if (config.getOptionalValue(PROP_PRODUCER_MAX_MESSAGE_SIZE, Integer.class).isPresent()) {
+            this.mqProducer.setMaxMessageSize(config.getValue(PROP_PRODUCER_MAX_MESSAGE_SIZE, Integer.class));
+        }
+
+        this.mqProducer.setLanguage(LanguageCode.JAVA);
+        startProducer();
+    }
+
+    private void startProducer() {
+        try {
+            this.mqProducer.start();
+            LOGGER.info("Consumer started...");
+        }
+        catch (MQClientException e) {
+            throw new DebeziumException(e);
+        }
+    }
+
+    private String createUniqInstance(String prefix) {
+        return prefix.concat("-").concat(UUID.randomUUID().toString());
+    }
+
+    @PreDestroy
+    void close() {
+        // Closed rocketmq producer
+        LOGGER.info("Consumer destroy...");
+        if (mqProducer != null) {
+            mqProducer.shutdown();
+        }
+    }
+
+    @Override
+    public void handleBatch(List<ChangeEvent<Object, Object>> records, DebeziumEngine.RecordCommitter<ChangeEvent<Object, Object>> committer)
+            throws InterruptedException {
+        final CountDownLatch latch = new CountDownLatch(records.size());
+        for (ChangeEvent<Object, Object> record : records) {
+            try {
+                LOGGER.trace("Received event '{}'", record);
+                final String topicName = streamNameMapper.map(record.destination());
+                mqProducer.send(new Message(topicName, null, getString(record.key()), getBytes(record.value())),
+                        new SendCallback() {
+                            @Override
+                            public void onSuccess(SendResult sendResult) {
+                                LOGGER.trace("Sent message with offset: {}", sendResult.getQueueOffset());
+                                latch.countDown();
+                            }
+
+                            @Override
+                            public void onException(Throwable throwable) {
+                                LOGGER.error("Failed to send record to {}:", record.destination(), throwable);
+                                throw new DebeziumException(throwable);
+                            }
+                        });
+                committer.markProcessed(record);
```

Comment:
> > Calling onSuccess() may cause the offset submission to be out of order
> 
> right, calling it in `onSuccess` may result into out of order offsets. However, calling it here is also wrong as it will mark as processed messages that are not delivered yet and the delivery may fail. So you probably need special loop over the records after `latch.await();` where you marked them as processed once you are sure that they were really delivered.

I have modified it